src/driver-marsgaming/marsgaming-buttons.c

```diff
@@ -0,0 +1,425 @@
+
+static struct marsgaming_button_info
+ratbag_to_marsgaming_action_button(struct ratbag_button *button)
+{
+	const unsigned int button_id = button->action.action.button;
+	const struct ratbag_to_marsgaming_button_map *map;
+	ARRAY_FOR_EACH(ratbag_to_marsgaming_button_maps, map) {
+		if (button_id == map->button_id)
+			return map->button_info;
+	}
+	return (struct marsgaming_button_info)MARSGAMING_BUTTON_ACTION_NONE;
+}
+
+struct ratbag_to_marsgaming_special_map {
+	enum ratbag_button_action_special special_id;
+	struct marsgaming_button_info button_info;
+};
+
+static const struct ratbag_to_marsgaming_special_map ratbag_to_marsgaming_special_maps[] = {
+	{ RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_CYCLE_UP, { .action = MARSGAMING_MM4_ACTION_DPI_SWITCH } },
+	{ RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_DOWN, { .action = MARSGAMING_MM4_ACTION_DPI_MINUS } },
+	{ RATBAG_BUTTON_ACTION_SPECIAL_RESOLUTION_UP, { .action = MARSGAMING_MM4_ACTION_DPI_PLUS } },
+	{ RATBAG_BUTTON_ACTION_SPECIAL_PROFILE_CYCLE_UP, { .action = MARSGAMING_MM4_ACTION_PROFILE_SWITCH } },
+};
+
+static struct marsgaming_button_info
+ratbag_to_marsgaming_action_special(struct ratbag_button *button)
+{
+	const enum ratbag_button_action_special special_type = button->action.action.special;
+	const struct ratbag_to_marsgaming_special_map *map;
+	ARRAY_FOR_EACH(ratbag_to_marsgaming_special_maps, map) {
+		if (special_type == map->special_id)
+			return map->button_info;
+	}
+	return (struct marsgaming_button_info)MARSGAMING_BUTTON_ACTION_NONE;
+}
+
+static int
+ratbag_action_keycodes_from_macro(struct ratbag_button_action *action, unsigned int *key0_out, unsigned int *key1_out, unsigned int *mods_out);
+
+static struct marsgaming_button_info
+ratbag_to_marsgaming_action_macro(struct ratbag_button *button)
+{
+	unsigned int key0;
+	unsigned int mods;
+	int e = ratbag_action_keycode_from_macro(&button->action, &key0, &mods);
+	if (e == 1) { // We can convert this to single/combo key
+		return (struct marsgaming_button_info){
+			.action = MARSGAMING_MM4_ACTION_COMBO_KEY,
+			.action_info.combo_key = {
+				.modifiers = mods,
+				.keys[0] = ratbag_hidraw_get_keyboard_usage_from_keycode(button->profile->device, key0),
+			}
+
+		};
+	}
+	unsigned int key1;
+	e = ratbag_action_keycodes_from_macro(&button->action, &key0, &key1, &mods);
+	if (e == 2) { // We can convert this to 2-key combo
+		return (struct marsgaming_button_info){
+			.action = MARSGAMING_MM4_ACTION_COMBO_KEY,
+			.action_info.combo_key = {
+				.modifiers = mods,
+				.keys[0] = ratbag_hidraw_get_keyboard_usage_from_keycode(button->profile->device, key0),
+				.keys[1] = ratbag_hidraw_get_keyboard_usage_from_keycode(button->profile->device, key1),
+			}
+
+		};
+	}
+	// TODO: Implement macros
+	return (struct marsgaming_button_info)MARSGAMING_BUTTON_ACTION_NONE;
+}
+
+static int
+ratbag_action_keycodes_from_macro(struct ratbag_button_action *action, unsigned int *key0_out, unsigned int *key1_out, unsigned int *modifiers_out)
```

Comment:
Note to self: take a look at this function again later.
Also, return value of this function is confusing. it would make sense if the output was written to an array, but here two separate arguments are used.

Reply:
I deleted this function, I'm resolving this conversation.